Add host VM - GPU compatibility checks for GCP

WoosukKwon opened the pull request:
This PR checks compatibility between GCP host VMs and accelerators. For example, GPUs (except A100) can be only attached to N1 machines, and each GPU has limitations on the number of vCPUs and amount of CPU memory that its host VM can have. This PR hard-codes such information in `gcp_catalog.py` and lets users know when their requests are invalid.

Tested:
* `sky gpunode --instance-type n1-highmem-16 --gpus K80 -c test` (invalid)
* `sky gpunode --instance-type n1-highmem-16 --gpus K80:2 -c test` (valid)
* `sky gpunode --instance-type n1-highmem-16 --gpus A10G -c test` (invalid)
* `sky gpunode --instance-type a2-highgpu-1g -c test` (invalid)
* `sky gpunode --instance-type a2-highgpu-1g --gpus A100:2 -c test` (invalid)
* `sky gpunode --instance-type n1-highcpu-16 --gpus A100 -c test` (invalid)

concretevitamin commented on tests/test_optimizer_random_dag.py:
```diff
@@ -65,8 +64,8 @@ def generate_random_dag(
                     cloud=CLOUDS[candidate.cloud],
                     instance_type=candidate.instance_type \
-                        if candidate.cloud != 'GCP' \
-                        else random.choice(GCP_INSTANCE_TYPES),
+                        if not pd.isna(candidate.instance_type) \
+                        else CLOUDS['GCP'].get_default_instance_type(),
```
Remove backslashes and quote in parens so yapf can auto-reformat?

WoosukKwon replied:
Fixed. For readability, I removed the set comprehension and added a loop statement instead.

concretevitamin commented on sky/clouds/service_catalog/__init__.py:
```diff
@@ -159,6 +159,15 @@ def get_region_zones_for_accelerators(
+                                         accelerators: Optional[Dict[str, int]],
+                                         zone: Optional[str] = None,
+                                         clouds: CloudFilter = None) -> None:
+    """Check if GCP host VM type is compatible with the accelerators."""
```
Add: `GCP only: Check ...`?

WoosukKwon replied:
Added.

concretevitamin commented on sky/clouds/service_catalog/gcp_catalog.py:
```diff
@@ -373,3 +406,87 @@ def get_region_zones_for_accelerators(
+        return
+    acc_name = acc_name.upper()
+
+    # Treat A100 as a special case.
+    if acc_name == 'A100':
+        # A100 must be attached to A2 instance type.
+        if not instance_type.startswith('a2-highgpu-'):
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(
+                    f'A100 GPUs cannot be attached to {instance_type}. '
+                    'Use A2 instance type instead.')
+        a100_instance_type = _A100_INSTANCE_TYPES[acc_count]
+        if instance_type != a100_instance_type:
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(
+                    f'A100:{acc_count} cannot be attached to {instance_type}. '
+                    f'Use {a100_instance_type} instead.')
+        return
```
Testing this on master:
```
» sky gpunode --gpus A100 -t a2-highgpu-1g
I 07-29 17:26:11 optimizer.py:605] == Optimizer ==
I 07-29 17:26:11 optimizer.py:628] Estimated cost: $3.7 / hour
I 07-29 17:26:11 optimizer.py:628]
I 07-29 17:26:11 optimizer.py:675] Considered resources (1 node):
I 07-29 17:26:11 optimizer.py:703] ------------------------------------------------------------
I 07-29 17:26:11 optimizer.py:703]  CLOUD   INSTANCE        ACCELERATORS   COST ($)   CHOSEN
I 07-29 17:26:11 optimizer.py:703] ------------------------------------------------------------
I 07-29 17:26:11 optimizer.py:703]  GCP     a2-highgpu-1g   A100:1         3.67          ✔
I 07-29 17:26:11 optimizer.py:703] ------------------------------------------------------------
I 07-29 17:26:11 optimizer.py:703]
Launching a new cluster 'sky-gpunode-zongheng'. Proceed? [Y/n]: n
Aborted!

» sky gpunode  -t a2-highgpu-1g                                                       1 ↵
I 07-29 17:26:20 optimizer.py:605] == Optimizer ==
I 07-29 17:26:20 optimizer.py:628] Estimated cost: $0.7 / hour
I 07-29 17:26:20 optimizer.py:628]
I 07-29 17:26:20 optimizer.py:675] Considered resources (1 node):
I 07-29 17:26:20 optimizer.py:703] ------------------------------------------------------------
I 07-29 17:26:20 optimizer.py:703]  CLOUD   INSTANCE        ACCELERATORS   COST ($)   CHOSEN
I 07-29 17:26:20 optimizer.py:703] ------------------------------------------------------------
I 07-29 17:26:20 optimizer.py:703]  GCP     a2-highgpu-1g   -              0.74          ✔
I 07-29 17:26:20 optimizer.py:703] ------------------------------------------------------------
I 07-29 17:26:20 optimizer.py:703]
Launching a new cluster 'sky-gpunode-zongheng'. Proceed? [Y/n]:
```
I thought A100s on GCP are special in that they come with the a2 VMs, and we should not allow the first case (attaching A100 to a2 VM) to happen?

WoosukKwon replied:
If I understand correctly, A100 GPUs must be attached to A2 VMs, just like other types of GPUs must be attached to N1 VMs. TPU nodes are the only exceptions that host VMs come with the accelerators.

concretevitamin replied:
I see. On master I could launch a no-GPU version of `a2-highgpu-1g`. On GCP console, there doesn't appear a way to do so -- one must select A100 GPUs first which will unlock the choice of an `a2` VM.  I think this PR's treatment of banning CPU-only A2 is good as it's consistent with the GCP console (not the API loophole).

concretevitamin commented on sky/clouds/service_catalog/gcp_catalog.py:
```diff
@@ -373,3 +406,87 @@ def get_region_zones_for_accelerators(
+    elif machine_type == 'highmem':
+        memory = 6.5 * num_cpus
+    elif machine_type == 'highcpu':
+        memory = 0.9 * num_cpus
```
Add 
```
else:
   raise ValueError(..)
```
?

WoosukKwon replied:
Added. In the future, we should get this information directly from the catalog.

concretevitamin commented on sky/clouds/service_catalog/gcp_catalog.py:
```diff
@@ -373,3 +406,87 @@ def get_region_zones_for_accelerators(
+        with ux_utils.print_exception_no_traceback():
+            raise ValueError(
+                f'{acc_name} GPUs cannot be attached to {instance_type}. '
+                f'Use N1 instance types instead.')
```
Shall we add L452 "Refer to: url" here too? Same comment for all other raise's in this func.

WoosukKwon replied:
Added.

concretevitamin reviewed (commented):
Very nice @WoosukKwon  - now users don't need to go through the failover loop observing the HttpError's. Some comments.

WoosukKwon commented:
@concretevitamin Thanks for your review! While I addressed all of your comments, I found that this PR breaks `sky exec` and `sky launch -c existing-cluster`. For existing clusters, we only need to check if the resource request is less demanding than what the cluster has. Thus, the `check_host_accelerator_availability` function should be called only when a new cluster is launched.

concretevitamin commented on sky/clouds/service_catalog/gcp_catalog.py:
```diff
@@ -248,3 +281,87 @@ def get_region_zones_for_accelerators(
+    # Check maximum vCPUs and memory.
+    if acc_name not in _NUM_ACC_TO_MAX_CPU_AND_MEMORY:
+        with ux_utils.print_exception_no_traceback():
+            raise ValueError(f'{acc_name} is not supported by GCP.')
```
Add: `See `sky show-gpus --cloud gcp`.`

WoosukKwon replied:
Added.

concretevitamin reviewed (approved):
Thanks!

WoosukKwon commented:
@concretevitamin I made the compatibility check invoked by the optimizer. Now this PR does not break `sky launch` and `sky exec` on existing clusters. However, a slight downside of this implementation is that in `sky spot launch` the compatibility check is not made until the spot controller runs the optimizer. I think we can address this in a future PR. PTAL.

concretevitamin commented on sky/clouds/service_catalog/gcp_catalog.py:
```diff
@@ -267,3 +301,93 @@ def get_region_zones_for_accelerators(
+                'Use N1 instance types instead. Please refer to: '
+                'https://cloud.google.com/compute/docs/machine-types#gpus')
+
+    # Check maximum vCPUs and memory.
+    if acc_name not in _NUM_ACC_TO_MAX_CPU_AND_MEMORY:
+        with ux_utils.print_exception_no_traceback():
+            raise exceptions.ResourcesUnavailableError(
+                f'{acc_name} is not available in GCP. '
+                'See \'sky show-gpus --cloud gcp\'')
```
Should this be `assert acc_name in _NUM_ACC_TO_MAX_CPU_AND_MEMORY`?

It should've been caught outside. E.g., under this branch
```
» sky launch --cloud gcp --gpus M60  ''                                                         1 ↵
I 08-29 15:02:52 optimizer.py:879] No resource satisfying {'M60': 1} on [GCP].
sky.exceptions.ResourcesUnavailableError: No launchable resource found for task sky-cmd. To fix: relax its resource requirements.
Hint: 'sky show-gpus --all' to list available accelerators.
      'sky check' to check the enabled clouds.
```

WoosukKwon replied:
That's a good point. Actually, `--cloud gcp --gpus M60 ''` and ` --instance-type n1-highmem-8 --gpus M60` will raise different error messages:

1.
```
$ sky launch --cloud gcp --gpus M60 ''
I 08-30 22:06:54 optimizer.py:875] No resource satisfying {'M60': 1} on [GCP].
sky.exceptions.ResourcesUnavailableError: No launchable resource found for task sky-cmd. To fix: relax its resource requirements.
Hint: 'sky show-gpus --all' to list available accelerators.
      'sky check' to check the enabled clouds.
```

2.
```
$ sky launch --instance-type n1-highmem-8 --gpus M60 ''
sky.exceptions.ResourcesUnavailableError: M60 is not available in GCP. See 'sky show-gpus --cloud gcp'
```

In the first case, the optimizer asks itself which instance to choose, and finds that GCP does not support M60. On the other hand, in the second case, the optimizer checks whether M60 can be attached to n1-highmem-8, and the new checks added in `gcp_catalog` finds the error. Since the two cases take different paths, the error messages are different.

concretevitamin reviewed (approved):
Nice @WoosukKwon! Consider running the smoke tests before merging.

WoosukKwon commented:
I changed the implementation substantially. The PR now consists of two new functions `check_host_accelerator_compatibility` and `check_accelerator_attachable_to_host`.

The first `check_host_accelerator_compatibility` function is invoked when `Resources` objects are created. It simply checks that accelerators are used with N1 machines, and does NOT check the maximum vCPU count and maximum memory limits for the accelerator because any Resources like `GCP(n1-highmem-64, {'V100': 0.01}` are allowed for `sky exec`.

The second `check_accelerator_attachable_to_host` function checks the cpu and memory limits. It is invoked by the optimizer, so `sky exec` will not execute this function.

@concretevitamin Could you please take another look?

concretevitamin commented on sky/optimizer.py:
```diff
@@ -887,4 +883,10 @@ def _fill_in_launchable_resources(
         launchable[resources] = _filter_out_blocked_launchable_resources(
             launchable[resources], blocked_launchable_resources)
 
+        for r in launchable[resources]:
```
Q: why move it to here, rather than after L849? Was thinking checking `resources` in that loop makes more sense, as it represents a validation of the user-requested resources.  Here, it may be possible than `launchable[resources]` has more than 1 "expanded" resources, and throwing an error on these may be unexpected?

concretevitamin replied:
I meant that here launchable[resources] may have more than 1 element, - can some of them pass the check, while some fail?  In these cases it may make sense to remove the candidates that fail rather than raising an error to the whole program.

WoosukKwon replied:
OK. That makes sense. I've rolled back the change.

concretevitamin commented on sky/clouds/service_catalog/__init__.py:
```diff
@@ -190,10 +190,26 @@ def get_region_zones_for_accelerators(
-                                         zone: Optional[str] = None,
                                          clouds: CloudFilter = None) -> None:
-    """GCP only: Check if host VM type is compatible with the accelerators."""
+    """GCP only: Check if host VM type is compatible with the accelerators.
```
Can we add https://github.com/skypilot-org/skypilot/pull/989#issuecomment-1232479501 to this func and the next func (L207+)? It's great explanation on why these two funcs are structured this way.

WoosukKwon replied:
Added.

concretevitamin reviewed (approved):
LGTM @WoosukKwon with a minor question. Reminder to rerun smoke tests before merging.

WoosukKwon commented:
I've checked that this PR does not break any smoke test.

WoosukKwon commented:
@concretevitamin If you don't have any more concern about this PR, I'll merge it.

concretevitamin commented:
Let’s ship it!

On Wed, Aug 31, 2022 at 14:31 Woosuk Kwon ***@***.***> wrote:

> @concretevitamin <https://github.com/concretevitamin> If you don't have
> any more concern about this PR, I'll merge it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/skypilot-org/skypilot/pull/989#issuecomment-1233441521>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEQWHUJYT5XIY3CKBCLAH3V37FJ5ANCNFSM533DAXVQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
